### PR TITLE
Upgrade stateful repo ops testing significantly

### LINKED ISSUE #1662: More verification when creating repo object
The methods that create the repo object in `repo_info.rs` should verify that we are creating a valid object. Examples:
* all branches point to existing snapshot_ids
* all tags point to existing snapshot_ids
* all parent snapshot_ids exist
* etc

Ideally we should do this somewhat efficiently, so for example, if we are creating a new branch we only check that snapshot. But this may not be necessary, and full verification would be safer. We should decide what's the right approach.

### PR DESCRIPTION
**Pull request by dcherian**
~Note this branch has squashed in #1708  so that should be merged first.~
~Note this branch has squashed in #1717  so that should be merged first.~

Closes #1662 

- explicitly models ops log
- reduce `num_updates` per config file
- increases ancestry generated
- fixes GC & expiration rules so that they actually work

### CODE REVIEW

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 1]
```diff
@@ -793,6 +825,26 @@ def check_ancestry(self) -> None:
...
+            raise NotImplementedError()
+
+        info = self.repo.inspect_repo_info()
+
+        assert info["spec_version"] == ver
+        assert set(info["deleted_tags"]) == self.model.deleted_tags
+        assert info["metadata"] == self.model.metadata
+
+        # TODO: something about latest_updates
+        # assert info["latest_updates"]
```
Comment (dcherian) [inline icechunk-python/tests/test_stateful_repo_ops.py]: Given how important it is, should I model the updates more explicitly and ensure `["latest updates"]` is always exactly right?

Comment (paraseba) [reply icechunk-python/tests/test_stateful_repo_ops.py]: Yes please, this is critical to the consistency of the repo. In particular, I'd force the config to have a small number of updates per file so we are also exercising the transition between files.

**Comment by dcherian** [comment]
Latest changes make it so that we build much longer history and so that expiration & GC are more effective.

Some statistics: `0-9` is a bin so that number of backup paths in a run

Test takes 30s on my machine

```
      * 100.00%, ancestry length: 0-9
      * 39.50%, backups exist
      * 31.09%, snapshots garbage collected: 0
      * 25.21%, branches expired: 0
      * 25.21%, commits expired: 0
      * 25.21%, tags expired: 0
      * 23.53%, ancestry length: 10-19
      * 17.65%, ancestry length: 20-29
      * 7.56%, branches expired: 1
      * 6.72%, commits expired: 1
      * 5.88%, ancestry length: 40-49
      * 5.88%, branches expired: 2
      * 5.88%, snapshots garbage collected: 1
      * 5.04%, tags expired: 1
      * 4.20%, ancestry length: 50-59
      * 3.36%, ancestry length: 60-69
      * 3.36%, commits expired: 10
      * 1.68%, snapshots garbage collected: 11
      * 0.84%, commits expired: 37
      * 0.84%, commits expired: 4
      * 0.84%, commits expired: 6
      * 0.84%, snapshots garbage collected: 37
      * 0.84%, snapshots garbage collected: 6
      * 0.84%, snapshots garbage collected: 8
```

File: icechunk-python/src/config.rs [hunk 2]
```diff
@@ -1957,4 +1958,35 @@ impl PyStorage {
         })?;
         Ok(res.into())
     }
+
+    #[pyo3(signature = (settings=None, prefix=None))]
+    pub(crate) fn list_objects(
```
Comment (dcherian) [inline icechunk-python/src/config.rs:1968]: so that i can test "file invariants"

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 3]
```diff
@@ -130,10 +224,22 @@ def __init__(self, **kwargs: Any) -> None:
         self.ondisk_snaps: dict[str, CommitModel] = {}
         self.tags: dict[str, TagModel] = {}
         self.branch_heads: dict[str, str] = {}
+        self.deleted_tags: set[str] = set()
+
+        self.ops_log: list[UpdateModel] = []
```
Comment (dcherian) [inline icechunk-python/tests/test_stateful_repo_ops.py:236]: explicitly models each op now

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 4]
```diff
@@ -792,11 +940,67 @@ def check_ancestry(self) -> None:
...
+
+        info = self.repo.inspect_repo_info()
+
+        assert info["spec_version"] == expected
+        assert set(info["deleted_tags"]) == self.model.deleted_tags
+        assert info["metadata"] == self.model.metadata
+
+        # the remaining fields (snapshots, branches, tags, ops_log) are checked by the other invariants
+
+    def check_file_invariants(self) -> None:
```
Comment (dcherian) [inline icechunk-python/tests/test_stateful_repo_ops.py:1015]: What else can I check?

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 5]
```diff
@@ -792,11 +940,67 @@ def check_ancestry(self) -> None:
...
+            for (first, second) in itertools.pairwise(actual_ops)
+        )
+        assert self.model.ops_log[::-1] == actual_ops
+        assert isinstance(actual_ops[-1], ic.RepoInitializedUpdate)
 
 
 VersionControlStateMachine.TestCase.settings = settings(
     deadline=None,
-    # stateful_step_count=100,
+    stateful_step_count=100,
```
Comment (dcherian) [inline icechunk-python/tests/test_stateful_repo_ops.py]: default is 50; i can remove this and make a "ci" or "nightly' profile that goes longer if you prefer

Comment (ianhi) [reply icechunk-python/tests/test_stateful_repo_ops.py]: I think 100 is probably fine for this. but I do think that a nightly version that goes crazy long (1000?) could be nice to have

File: icechunk-python/src/config.rs [hunk 6]
```diff
@@ -1957,4 +1958,35 @@ impl PyStorage {
...
+        let prefix = prefix.unwrap_or_default();
+        let settings: Option<storage::Settings> = settings.map(|s| s.into());
+        pyo3_async_runtimes::tokio::get_runtime().block_on(async {
+            let settings = match settings {
+                Some(s) => s,
+                None => self
+                    .0
+                    .default_settings()
+                    .await
+                    .map_err(|e| PyValueError::new_err(e.to_string()))?,
```
Comment (ianhi) [inline icechunk-python/src/config.rs]: this should probably be a merge in case a settings object with some null values is passed. don't want to repeat https://github.com/earth-mover/icechunk/issues/1689

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 7]
```diff
@@ -179,12 +285,23 @@ def commit(self, snap: SnapshotInfo) -> None:
...
+        assert self.branch is not None
+        self.ops_log.append(NewCommitUpdateModel(self.branch, snap.id))
 
     def amend(self, snap: SnapshotInfo) -> None:
         """Amend the HEAD commit."""
         # this is simple because we aren't modeling the branch as a list of commits
-        self.commit(snap)
+        assert self.branch is not None
+        prev_snap_id = self.branch_heads[self.branch]
+        self._commit(snap)
```
Comment (ianhi) [inline icechunk-python/tests/test_stateful_repo_ops.py]: Is it ok that that the branch assert is on different sides of the `_commit` call for commit vs amend?

Comment (ianhi) [reply icechunk-python/tests/test_stateful_repo_ops.py]: ahh this is in the mode. the github UI fooled me. So i'm guessing yes.

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 8]
```diff
@@ -610,29 +749,50 @@ def delete_branch(self, branch: str) -> None:
             with pytest.raises(IcechunkError):
                 self.repo.delete_branch(branch)
 
+    @rule(tag=consumes(tags))
+    def delete_tag(self, tag: str) -> None:
+        note(f"Deleting tag {tag!r}")
+        if tag not in self.model.tags:
+            with pytest.raises(IcechunkError):
+                self.repo.delete_tag(tag)
```
Comment (ianhi) [inline icechunk-python/tests/test_stateful_repo_ops.py:806]: Shouldn't this be an error? If we get here we must've done something wrong in our modelling?

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 9]
```diff
@@ -610,29 +749,50 @@ def delete_branch(self, branch: str) -> None:
...
+            self.repo.delete_tag(tag)
+            self.model.delete_tag(tag)
+
+    def _draw_older_than(self, data: st.DataObject) -> datetime.datetime:
+        """Draw an `older_than` timestamp biased toward actually expiring commits.
+
+        We draw from three buckets:
+        - future: max(commit_times) + 1 s  → all commits are eligible
+        - boundary: an exact commit time   → tests the strict-< boundary
+        - past: year-2000 sentinel         → nothing should be expired
```
Comment (ianhi) [inline icechunk-python/tests/test_stateful_repo_ops.py]: This is great

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 10]
```diff
@@ -687,19 +851,10 @@ def expire_snapshots(
             self.maybe_checkout_branch(branch)
 
     @precondition(lambda self: bool(self.model.commit_times))
-    @rule(
-        data=st.data(),
-        # we delete based on snapshot created_at time, not flushed_at time
-        # so for delta we just past an integer number of seconds to handle the
-        # (small) difference in tests
-        delta=st.integers(min_value=-86400, max_value=86400).filter(lambda x: x != 0),
```
Comment (ianhi) [inline icechunk-python/tests/test_stateful_repo_ops.py:725]: does this delta around the flused vs created times not matter now? Although reading this I'm not really sure what it was getting at .

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 11]
```diff
@@ -792,11 +940,68 @@ def check_ancestry(self) -> None:
...
+    def check_file_invariants(self) -> None:
+        assert self.storage is not None
+        paths = set(path_and_bytes[0] for path_and_bytes in self.storage.list_objects())
+
+        nsnapshots = len([p for p in paths if p.startswith("snapshots/")])
+        ntransactions = len([p for p in paths if p.startswith("transactions/")])
+        if self.model.spec_version == 1:
+            assert nsnapshots - 1 == ntransactions
+        else:
+            assert nsnapshots == ntransactions
```
Comment (ianhi) [inline icechunk-python/tests/test_stateful_repo_ops.py]: Pretty sure that a spec version 1 that was initially written by icechunk python 2 should have a transaction for the first snapshot now after https://github.com/earth-mover/icechunk/pull/1623

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 12]
```diff
@@ -792,11 +940,68 @@ def check_ancestry(self) -> None:
...
+
+        if self.model.spec_version > 1:
```
Comment (ianhi) [inline icechunk-python/tests/test_stateful_repo_ops.py]:
```suggestion
        if self.model.spec_version >= 2:
```

just to be consistent with the rest of the file.

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 13]
```diff
@@ -799,27 +959,89 @@ def check_branches(self) -> None:
...
+        if self.model.spec_version == 1:
+            return
+        actual_ops = list(self.repo.ops_log())
+        assert all(
+            first.updated_at > second.updated_at
+            for (first, second) in itertools.pairwise(actual_ops)
+        )
+        assert all(
+            first.backup_path != second.backup_path
+            for (first, second) in itertools.pairwise(actual_ops)
```
Comment (ianhi) [inline icechunk-python/tests/test_stateful_repo_ops.py]: can we be more specific here. shouldn't updates within the same file share part of the backup path?

File: icechunk-python/src/config.rs (hunk 2 above)
Comment (ianhi) [reply icechunk-python/src/config.rs:1968]: this is good. I also suspect it will be useful to users down the line. A docstring (even just on the python side) would be fantastic

File: icechunk-python/tests/test_stateful_repo_ops.py (hunk 13 above)
Comment (ianhi) [reply icechunk-python/tests/test_stateful_repo_ops.py]: ok wrote a little test script I think we can split on `.` and do some further testing of paths in the same file vs not:
```
  [16:57:21.744166] NewCommitUpdate
    branch: feature
    new_snap_id: D76R5BCP1THQ...
    backup_path: None

  [16:57:21.741749] TagDeletedUpdate
    name: v1
    previous_snap_id: M3AWG53ADW49...
    backup_path: repo.30731727758256.QSZ36J8N8ZKCX1TPY1KG

  [16:57:21.741122] TagCreatedUpdate
    name: v1
    backup_path: repo.30731727758259.WZK83RFKAWJTBBEFHWW0

  [16:57:21.740792] BranchCreatedUpdate
    name: feature
    backup_path: repo.30731727758259.ZNJMSQKQXD90GXWT5JRG
```

File: icechunk-python/tests/test_stateful_repo_ops.py (hunk 7 above)
Comment (dcherian) [reply icechunk-python/tests/test_stateful_repo_ops.py]: it's purely for type checking.

File: icechunk-python/tests/test_stateful_repo_ops.py (hunk 8 above)
Comment (dcherian) [reply icechunk-python/tests/test_stateful_repo_ops.py:806]: > with pytest.raises(IcechunkError):

It is an error...

Comment (ianhi) [reply icechunk-python/tests/test_stateful_repo_ops.py:806]: no i mean it should fail the test. This is catching thath icechunk has an error, but it masks a test bug where we have divergent tags bundle from the actual

File: icechunk-python/tests/test_stateful_repo_ops.py (hunk 10 above)
Comment (dcherian) [reply icechunk-python/tests/test_stateful_repo_ops.py:725]: IDK. it's broken now so I"m disabling. Will come back to this.

**Comment by ianhi** [comment]
> Some statistics: 0-9 is a bin so that number of backup paths in a run

> Test takes 30s on my machine

not sure i fully get this output. Also some of them are just 0 instead of bin?

**Comment by dcherian** [comment]
Did you run with `--hypothesis-show-statistics`?

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 14]
```diff
@@ -407,6 +525,23 @@ def new_store(self) -> None:
     def sync_store(self) -> NewSyncStoreWrapper:
         return NewSyncStoreWrapper(self.session.store)
 
+    @precondition(lambda self: self.model.branch is not None)
+    @rule(data=st.data(), ncommits=st.integers(3, 10), target=commits)
+    def set_and_commit(self, data: st.DataObject, ncommits: int) -> Any:
+        # We have so many rules now it is hard to build up long history
+        # instead we explicitly add a rule to make a lot of commits
```
Comment (paraseba) [inline icechunk-python/tests/test_stateful_repo_ops.py:546]: awesome idea, wonder if instead of ncommits you could get a list of booleans that tell you if you should do commit or amend for each

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 15]
```diff
@@ -799,27 +956,87 @@ def check_branches(self) -> None:
...
+        elif ver == 2:
+            expected = "2.0"
+        else:
+            raise NotImplementedError()
+
+        info = self.repo.inspect_repo_info()
+
+        assert info["spec_version"] == expected
+        assert set(info["deleted_tags"]) == self.model.deleted_tags
+        assert info["metadata"] == self.model.metadata
```
Comment (paraseba) [inline icechunk-python/tests/test_stateful_repo_ops.py:1011]: Should we add that there is agreement between ancestry and ops_log? For each NewCommit in ancestry before any expiration, there should be one and only one element in ancestry that agrees with it

File: icechunk-python/tests/test_stateful_repo_ops.py [hunk 16]
```diff
@@ -799,27 +956,87 @@ def check_branches(self) -> None:
...
+
+    def check_file_invariants(self) -> None:
+        if Version(self.ic.__version__).major == 1:
+            return
+
+        assert self.storage is not None
+        paths = set(path_and_bytes[0] for path_and_bytes in self.storage.list_objects())
+
+        nsnapshots = len([p for p in paths if p.startswith("snapshots/")])
+        ntransactions = len([p for p in paths if p.startswith("transactions/")])
```
Comment (paraseba) [inline icechunk-python/tests/test_stateful_repo_ops.py]: you could match snapshots id to tx ids, they should be 1 to 1

File: icechunk-python/tests/test_stateful_repo_ops.py (hunk 5 above)
Comment (paraseba) [reply icechunk-python/tests/test_stateful_repo_ops.py]: yeah, let's make a big one to runs periodically

Comment (paraseba) [reply icechunk-python/tests/test_stateful_repo_ops.py]: if it's easy...

File: icechunk-python/tests/test_stateful_repo_ops.py (hunk 8 above)
Comment (dcherian) [reply icechunk-python/tests/test_stateful_repo_ops.py:806]: That is intentional:

https://github.com/earth-mover/icechunk/blob/d45408d8c0e9d2fedabd492078d8172bea3c2a3b/icechunk-python/tests/test_stateful_repo_ops.py#L563-L579

File: icechunk-python/tests/test_stateful_repo_ops.py (hunk 10 above)
Comment (dcherian) [reply icechunk-python/tests/test_stateful_repo_ops.py:725]: Fixed now

File: icechunk-python/tests/test_stateful_repo_ops.py (hunk 13 above)
Comment (dcherian) [reply icechunk-python/tests/test_stateful_repo_ops.py]: i don't understand?

File: icechunk-python/tests/test_stateful_repo_ops.py (hunk 15 above)
Comment (dcherian) [reply icechunk-python/tests/test_stateful_repo_ops.py:1011]: Seems a bit hard to check. reset branch is another edge case + we'd need to search through ancestry of all branches & tags

Comment (paraseba) [reply icechunk-python/tests/test_stateful_repo_ops.py:1011]: yeah ... feel free to ignore
